### PR TITLE
[ExoBundle] Generic feedbacks

### PR DESCRIPTION
**Pull request by eldoniel**
This adds a few features to the formative exercises, concerning the steps' max attempts.

- Generic feedbacks are added to the questions. They are indications that the user found all answers or not.
- Suite feedbacks are added to the steps. They indicate to the user that he can try again, or not.
- Navigation is handled with consideration of these parameters.

For now, the maxAttempts and all the feedback sentences are hard coded.

### CODE REVIEW

**Comment by LaurentGruber** [comment]
LGTM

File: plugin/exo/Resources/public/js/angular/Exercise/Partials/player.html [hunk 1]
```diff
@@ -103,10 +105,10 @@
         <!-- End Exercise -->
         <button type="button" role="button" class="btn btn-primary"
                 data-ng-if="!exercisePlayerCtrl.next"
-                data-ng-disabled="!exercisePlayerCtrl.feedback.visible && exercisePlayerCtrl.feedback.enabled || (exercisePlayerCtrl.feedback.enabled && exercisePlayerCtrl.feedback.visible && !exercisePlayerCtrl.solutionShown)"
+                data-ng-disabled="exercisePlayerCtrl.isButtonEnabled('end')"
                 ng-bootbox-confirm="{{ 'sequence_end_sequence_confirm'|trans:{}:'ujm_sequence'}}"
                 ng-bootbox-confirm-action="exercisePlayerCtrl.end()"
-        >
+        ><!-- REVOIR LES CONDITIONS POUR LES BOUTONS. FAIRE UNE FONCTION SPÉCIFIQUE POUR GÉRER LES CONDITIONS? -->
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Exercise/Partials/player.html]: Isn't that the purpose of your function `isButtonEnabled` ? If it is, please remove comment.

File: plugin/exo/Resources/public/js/angular/Question/Controllers/Type/OpenQuestionCtrl.js [hunk 2]
```diff
@@ -39,17 +46,28 @@ OpenQuestionCtrl.prototype.onFeedbackShow = function onFeedbackShow() {
...
-                }
-            }
-        }
+		            this.answerWithKeywords = this.answerWithKeywords.replace(searchExpression, keyword, searchFlags);
+		        }
+		    }
+		}
+    }
+    
+    if (this.question.solutions.length === numAnswersFound) {
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Question/Controllers/Type/OpenQuestionCtrl.js]: this if block seems to be a duplicate of `OpenQuestionService.prototype.answersAllFound`.

File: plugin/exo/Resources/public/js/angular/Question/Services/Type/ChoiceQuestionService.js [hunk 3]
```diff
@@ -20,6 +20,39 @@ ChoiceQuestionService.prototype.initAnswer = function initAnswer() {
...
+        for (var j=0; j<answer.length; j++) {
+            if (question.solutions[i].id === answer[j] && question.solutions[i].score > 0) {
+                numAnswersFound++;
+            }
+        }
+    }
+    var feedbackState = -1;
+    if (numAnswersFound === numSolutions) {
+        // all answers have been found
+        feedbackState = 0;
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Question/Services/Type/ChoiceQuestionService.js]: inject FeedbackService and use class constants instead (like FeedbackService.MULTIPLE_ANSWERS_MISSING)

File: plugin/exo/Resources/public/js/angular/Exercise/Partials/player.html (hunk 1 above)
Comment (eldoniel) [inline plugin/exo/Resources/public/js/angular/Exercise/Partials/player.html]: It is indeed, thanks for pointing it out. Comment removed.

File: plugin/exo/Resources/public/js/angular/Question/Services/Type/ClozeQuestionService.js [hunk 4]
```diff
@@ -19,6 +19,44 @@ ClozeQuestionService.prototype.initAnswer = function initAnswer() {
...
+                    numAnswersFound++;
+                }
+            }
+        }
+    }
+    
+    var feedbackState = -1;
+    if (numAnswersFound === question.solutions.length) {
+        // all answers have been found
+        feedbackState = 0;
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Question/Services/Type/ClozeQuestionService.js]: idem inject FeedbackService and use class constants instead (like FeedbackService.MULTIPLE_ANSWERS_MISSING)

File: plugin/exo/Resources/public/js/angular/Question/Services/Type/GraphicQuestionService.js [hunk 5]
```diff
@@ -35,6 +35,61 @@ GraphicQuestionService.prototype.getCrosshair = function getCrosshair() {
...
+                || (question.solutions[i].shape === "square" && answerX > startX && answerX < endX && answerY > startY && answerY < endY))
+                && notFoundZones.indexOf(question.solutions[i]) !== -1) {
+                notFoundZones.splice(notFoundZones.indexOf(question.solutions[i]), 1);
+            }
+        }
+    }
+    
+    var feedbackState = -1;
+    if (notFoundZones.length === 0) {
+        feedbackState = 0;
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Question/Services/Type/GraphicQuestionService.js]: idem inject FeedbackService and use class constants instead (like FeedbackService.MULTIPLE_ANSWERS_MISSING)

File: plugin/exo/Resources/public/js/angular/Question/Services/Type/MatchQuestionService.js [hunk 6]
```diff
@@ -30,6 +30,39 @@ MatchQuestionService.prototype.getCorrectAnswer = function getCorrectAnswer(ques
...
+            if (question.solutions[j].firstId === answer[0] && question.solutions[j].secondId === answer[1]) {
+                numAnswersFound++;
+            }
+        }
+    }
+    
+    var feedbackState = -1;
+    if (numAnswersFound === question.solutions.length) {
+        // all answers have been found
+        feedbackState = 0;
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Question/Services/Type/MatchQuestionService.js]: idem inject FeedbackService and use class constants instead (like FeedbackService.MULTIPLE_ANSWERS_MISSING)

File: plugin/exo/Resources/public/js/angular/Question/Services/Type/OpenQuestionService.js [hunk 7]
```diff
@@ -20,6 +20,57 @@ OpenQuestionService.prototype.initAnswer = function initAnswer() {
...
+            }
+        }
+    }
+    else {
+        feedbackState = 0;
+    }
+    
+    var feedbackState = -1;
+    if (question.solutions.length === numAnswersFound) {
+        feedbackState = 0;
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Question/Services/Type/OpenQuestionService.js]: idem inject FeedbackService and use class constants instead (like FeedbackService.MULTIPLE_ANSWERS_MISSING)

File: plugin/exo/Resources/public/js/angular/Step/Controllers/StepShowCtrl.js [hunk 8]
```diff
@@ -37,6 +50,15 @@ StepShowCtrl.prototype.stepIndex = 0;
  */
 StepShowCtrl.prototype.solutionShown = false;
 
+StepShowCtrl.prototype.ALL_ANSWERS_FOUND = 0;
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Step/Controllers/StepShowCtrl.js]: reuse FeedbackService class constants instead (like FeedbackService.MULTIPLE_ANSWERS_MISSING)

File: plugin/exo/Resources/public/js/angular/Step/Directives/StepShowDirective.js [hunk 9]
```diff
@@ -11,10 +11,12 @@ var StepShowDirective = function StepShowDirective() {
...
         scope: {
-            step          : '=',
-            stepIndex     : '@',
-            currentTry    : '=',
-            solutionShown : '='
+            step            : '=',
+            stepIndex       : '@',
+            currentTry      : '=',
+            solutionShown   : '=',
+            feedback        : '=',
```
Comment (Elorfin) [inline plugin/exo/Resources/public/js/angular/Step/Directives/StepShowDirective.js]: no need to pass feedback object through directive parameters. Retrieve it from the FeedbackService.

**Comment by eldoniel** [comment]
@claroline/maintainers Latest commits pushed. It fixes all the small issues @Elorfin pointed out.

**Comment by stefk** [comment]
@claroline/ujm, is it ok to hardcode max step attempts to 5? 

@eldoniel, `else` and `else if` shouldn't be on a new line. Use:

```js
if {

} else {

}
```
Instead of:
```js
if {

}
else {

}
```

**Comment by eldoniel** [comment]
@stefk The step's maxAttempts is not meant to stay fixed, it just should be until we can change it in the editor. The number of 5 has been decided by JJQ.
UJM shall confirm, but it seemed to me that the step's maxAttempts was not used until I did it (unlike the exercise's maxAttempts, which was already used), so it should not be a problem.

**Comment by stefk** [comment]
LGTM

**Pull request merged**